src/twisted/internet/tcp.py

```diff
@@ -949,24 +949,6 @@ def reserve():
         """
 
 
-    def __enter__():
-        """
-        Release the underlying file descriptor so that code within the
-        context manager can open a new file.
-        """
-
-
-    def __exit__(excType, excValue, traceback):
```

Comment:
I think that this should not be removed. The enter and exit method are required for the context manager behaviour.

```
Interface      -> def __exit__(      excType, excValue, traceback):
Implementation -> def __exit__(self, excValue, excType, traceback):
```

The interface is correct... the order is `type, value, traceback`.
The implementation should be fixed.

Here is the documentation https://docs.python.org/3/reference/datamodel.html#object.__enter__

```python
object.__exit__(self, exc_type, exc_value, traceback)
```

Reply:
That didn't work.  I changed to:

``
diff --git a/src/twisted/internet/tcp.py b/src/twisted/internet/tcp.py
index 0f493c091..e8f955317 100644
--- a/src/twisted/internet/tcp.py
+++ b/src/twisted/internet/tcp.py
@@ -956,7 +956,7 @@ class _IFileDescriptorReservation(Interface):
         """


-    def __exit__(excType, excValue, traceback):
+    def __exit__(exc_type, exc_value, traceback):
         """
         Attempt to re-open the reserved file descriptor.  See
         L{reserve} for caveats.
@@ -1024,7 +1024,7 @@ class _FileDescriptorReservation(object):
         self._fileDescriptor = None


-    def __exit__(self, excValue, excType, traceback):
+    def __exit__(self, exc_value, exc_type, traceback):
         """
         See L{_IFileDescriptorReservation.__exit__}.
         """
@@ -1066,7 +1066,7 @@ class _NullFileDescriptorReservation(object):
         """


-    def __exit__(self, excValue, excType, traceback):
+    def __exit__(self, exc_value, exc_type, traceback):
         """
         Do nothing.  See L{_IFileDescriptorReservation.__exit__}.
```

and still got:

```
src/twisted/internet/tcp.py:1027:5: error: Signature of "_FileDescriptorReservation" incompatible with "__exit__" of supertype "_IFileDescriptorReservation"  [override]
src/twisted/internet/tcp.py:1069:5: error: Signature of "_NullFileDescriptorReservation" incompatible with "__exit__" of supertype "_IFileDescriptorReservation"  [override]
```

Reply:
I have applied the following patch on top of `f541fa566 Craig Rodrigues 2020-07-02 Remove __enter__() and __exit__() from  _IFileDescriptorReservation
` and it worked

```diff
diff --git a/src/twisted/internet/tcp.py b/src/twisted/internet/tcp.py
index 79c049a1a..b69119a39 100644
--- a/src/twisted/internet/tcp.py
+++ b/src/twisted/internet/tcp.py
@@ -1006,7 +1006,7 @@ class _FileDescriptorReservation(object):
         self._fileDescriptor = None
 
 
-    def __exit__(self, excValue, excType, traceback):
+    def __exit__(self, excType, excValue, traceback):
         """
         See L{_IFileDescriptorReservation.__exit__}.
         """
```

Reply:
Oh OK, this worked:

```
--- a/src/twisted/internet/tcp.py
+++ b/src/twisted/internet/tcp.py
@@ -1024,7 +1024,7 @@ class _FileDescriptorReservation(object):
         self._fileDescriptor = None


-    def __exit__(self, excValue, excType, traceback):
+    def __exit__(self, excType, excValue, traceback):
         """
         See L{_IFileDescriptorReservation.__exit__}.
         """
@@ -1066,7 +1066,7 @@ class _NullFileDescriptorReservation(object):
         """


-    def __exit__(self, excValue, excType, traceback):
+    def __exit__(self, excType, excValue, traceback):
         """
         Do nothing.  See L{_IFileDescriptorReservation.__exit__}.
```

Reply:
Fixed